n3fit/src/n3fit/model_trainer.py

```diff
@@ -327,15 +326,39 @@ def _model_generation(self, pdf_models, partition, partition_idx):
         """
         log.info("Generating the Model")
 
-        # Construct the input array that will be given to the pdf
-        input_arr = np.concatenate(self.input_list, axis=1).T
+        # In the case of pineappl models all fktables ask for the same grid in x
+        # and so the input can be simplified to be a single grid for all dataset
+        # instead of a concatenation that gets splitted afterwards
+        # However, this is not a _strict_ requirement for pineappl so the solution below
+        # aims to be completely general
+        # Detailed:
+        #    let's assume an input [x1, x1, x1, x2, x2, x3]
+        #    where each xi is a different grid, this will be broken into two lists:
+        #    [x1, x2, x3] (unique grids) and [0,0,0,1,1,2] (index of the grid per dataset)
+        #    The pdf will then be evaluated to concatenate([x1,x2,x3]) and then split (x1, x2, x3)
+        #    Then each of the experiment, looking at the indexes, will receive one of the 3 PDFs
+
+        # Clean the input list
+        inputs_hash = []
+        inputs_unique = []
+        inputs_idx = []
+        for input_grid in self.input_list:
+            ghash = hash(tuple(input_grid.flatten()))
+            if ghash not in inputs_hash:
+                inputs_hash.append(ghash)
+                inputs_unique.append(input_grid)
+            inputs_idx.append(inputs_hash.index(ghash))
```

Comment:
I am not feeling brave enough to fully understand this logic right now, but seems to me this approach is flacky and complicated. First of all hashes are not necessarily unique, but also computing it seems unneeded given that is a big expense to do only once and the right interface is already provided by dicts and sets.  Putting something like `tuple(input_grid.flatten())` in a dict would be correct, faster (dicts can avoid computing hashes for the same pointer) and avoid the O(N) list lookup.

Reply:
I'm not entirely sure what you mean by the right interface. A set doesn't really work here (I need the order both outside and inside the arrays!) and a dictionary seems to me more complicated than plain lists.

I guess I can do this and remove the hash if you like it better.
```python
inputs_unique = []
inputs_unique_tuple = []
inputs_idx = []
for input_grid in self.input_list:
    tarr = tuple(input_grid.flatten())
    if tarr not in inputs_unique_tuple:
        inputs_unique.append(input_grid)
        inputs_unique_tuple.append(tarr)
    inputs_idx.append(inputs_unique_tuple.index(tarr))
```

Also, an O(N) list lookup is very much not a problem here. If it were this would not be needed in the first place. 

Actually I'm wondering now whether I shouldn't, instead of doing the `in` compare the arrays using `allclose` with some threshold.  I would not want to compute the grid twice just because in one you have `0.0000133333` and in the other `0.0000133334`...

Reply:
The two straight forward way would be to put the frozen versions of the inputs in a set/dictionary (which you were somehow doing minus the dictionary) or to compare manually the items. I guess your approach below would be sensible if you do want the allclose checks. 

I also wonder if it would make sense to "name" the grids somewhere so there is a name associated with a grid and we match by name (as in `pinneaple.default_fkgrid` or somesuch). This might help the toolchain enforce that the grids are actually the same, and require a mindful decision to do anything else. As a counterpoint, negative experience with reportengine suggests that just hashing the values would have been better than the complicated namings there (though I don't expect this to face the same problems).

Reply:
Sadly forcing all of them to be the same might not be the best because at the very least there is integrability which is a very special grid so we would have to have at least one exception.

Rethinking about it, the `allclose` is the best thing because as @RoyStegeman `n3fit` should just take whatever information receives (from pineappl in this case) and if the grids are close they produce the same pdf otherwise they don't.